osgbuild/fetch_sources.py

```diff
@@ -32,194 +82,263 @@
+    _, rc = utils.sbacktick(['git', 'rev-parse', '-q', '--verify', spec_rev])
+    if rc:
+        log.debug("No spec file found under %s" % spec_rev)
+        return None
+    with open(dest_spec, "w") as specf:
+        utils.checked_call(['git', 'show', spec_rev], stdout=specf)
+    return dest_spec
+
+def check_git_hash(url, tag, sha, got_sha, nocheck):
+    efmt = "Hash mismatch for %s tag %s\n    expected: %s\n    actual:   %s"
+    if sha != got_sha and deref_git_sha(sha) != deref_git_sha(got_sha):
+        msg = efmt % (url, tag, sha, got_sha)
+        if nocheck:
+            log.warning(msg + "\n    (ignored)")
+        else:
+            raise Error(msg)
 
 def deref_git_sha(sha):
-    output, rc = utils.sbacktick(["git", "rev-parse", sha + "^{}"])
+    cmd = ["git", "rev-parse", "-q", "--verify", sha + "^{}"]
+    output, rc = utils.sbacktick(cmd)
     if rc:
-        raise Error("Git failed to parse rev: '%s'" % sha)
+        log.error("Git failed to parse rev: '%s'" % sha)
+        return sha
     return output
 
-def process_dot_source(cache_prefix, sfilename, destdir, nocheck):
-    """Read a .source file, fetch any files mentioned in it from the
-    cache.
+def process_source_line(line, ops):
+    args,kv = parse_source_line(line)
+
+    handlers = dict(
+        git    = fetch_git_source,
+        github = fetch_github_source,
+        cached = fetch_cached_source,
+        uri    = fetch_uri_source,
+    )
+    explicit_type = kv.pop('type', None)
+    meta_type = explicit_type or get_auto_source_type(*args, **kv)
+    if meta_type in handlers:
+        handler = handlers[meta_type]
+        try:
+            return handler(*args, ops=ops, **kv)
+        except TypeError as e:
+            fancy_source_error(meta_type, explicit_type, handler, args, kv, e)
+    else:
+        raise Error("Unrecognized type '%s' (valid types are: %s)"
+                    % (meta_type, sorted(handlers)))
+
+def get_auto_source_type(*args, **kw):
+    if not args:
+        raise Error("No type specified and no default arg provided")
+    if args[0].endswith('.git'):
+        raise Error("No automatic types allowed for git sources")
+    if re.search(r'^\w+://', args[0]) or args[0].startswith('/'):
+        return 'uri'
+    else:
+        return 'cached'
+
+def parse_source_line(line):
+    kv, args = dual_filter((lambda t: t[0]), map(kvmatch, line.split()))
+    return [ a[1] for a in args ], dict(kv)
+
+def dual_filter(cond, seq):
+    pos,neg = [],[]
+    for x in seq:
+        (pos if cond(x) else neg).append(x)
+    return pos,neg
+
+def kvmatch(arg):
+    # return (key,val) for "key=val", else return (None, arg)
+    return re.search(r'^(?:(\w+)=)?(.*)', arg).groups()
+
+def fancy_source_error(meta_type, explicit_type, handler, args, kw, e):
+    xtype = "type" if explicit_type else "implicit type"
+    log.error("Error processing source line of %s '%s'" % (xtype, meta_type))
+    varnames = handler.__code__.co_varnames
```

Comment:
I have no idea what any of this is or how it works.

Reply:
>I have no idea what any of this is or how it works

that's why i call it, _fancy_

;)

Reply:
so... can you explain? In comments?